### DISCUSSION TITLE
Queue index file cannot be recovered if it has messages with TTL on 3.8.17+

### DISCUSSION THREAD

**Discussion by annieblomgren** [rabbitmq/rabbitmq-server#3272] [Questions] [answered]
Hi,

We've had two different clusters upgrading to 3.8.19 where the message store got corrupted:

1. First case is a single node with Erlang 23.3.1.
After the upgrades a queue got into the state 'down'. 

2. Second case is a 3 node cluster with Erlang 24.0.2.
A whole vhost was down.
We used strace to find the "faulty" .idx file and removed it, after that the vhost could start so not all messages were lost. 

We can share full logs and if of interest the data folder privately.

**Comment by johanrhodin** [comment]
More here: https://github.com/rabbitmq/rabbitmq-server/pull/3253#discussion_r681828930
and in this rabbitmq-users post: https://groups.google.com/g/rabbitmq-users/c/ekV9tTBRZms/m/kzZKsITSAwAJ

**Comment by annieblomgren** [comment]
It has happened again to a single node cluster running Erlang 24.0.4.
A common thing for all 3 cases so far is that the faulty queue has used TTL.

**Comment by carlhoerberg** [comment]
We at CloudAMQP have stopped provisioning 3.8.19 because so many customers gets corrupted messages stores from this version :warning:

**Comment by michaelklishin** [comment]
Most recent queue index changes were #2954 and #3041 in `3.8.17`.

**Comment by michaelklishin** [comment]
#2954 seems to be a lot more relevant than #3041. As of #2954, index files by default store 2048 entries instead of 16384. This value
is stored in the virtual host directory as `.config`. Existing installations should have the original value stored there. New installations will
use `rabbit.queue_index_segment_entry_count` or `2048` if it's not set. Setting `rabbit.queue_index_segment_entry_count` to `16384` should effectively undo #2954 for **newly created virtual hosts**.

Nodes will log

```
Setting segment_entry_count for vhost … with … queues to …
```

on boot and the `.config` file is a single Erlang term right now, and can easily be inspected.

Without knowing if this was an upgrade and from what version, we can't tell what segment entry count value should be in effect. But you
can inspect those values both using `rabbitmq-diagnostics environment` and by inspecting the dot files files under the virtual host directory (which would be under the message store root).

**Reply by michaelklishin** [reply]
Worth mentioning that https://github.com/rabbitmq/rabbitmq-server/pull/3068 was a follow-up PR that makes us back to the original segment count value when virtual host config file does not exist.

**Comment by annieblomgren** [comment]
Have been able to reproduce it in versions 3.8.19, 3.8.18 and 3.8.17. 
It works in 3.8.16.

To reproduce:
1. perftest command `bin/runjava com.rabbitmq.perf.PerfTest -x 1 -y 0 -u "throughput-test-1" -a --id "test 1" --message-properties expiration=1 -H $URL -f persistent -s 8000`
2. Restart RabbitMQ

Setting segment_entry_count for vhost 'X' with 0 queues to '2048'

**Reply by michaelklishin** [reply]
@annieblomgren thanks for the details, so this is with a brand new node and not an upgraded one?

**Reply by michaelklishin** [reply]
so this also includes per-message TTL, which makes me wonder if #3041 can be just as relevant.

**Reply by annieblomgren** [reply]
@michaelklishin yes tested on brand new nodes too

**Comment by michaelklishin** [comment]
We could reproduce and so far it seems that #3041 is the root cause. Reverting it seems to reliably make the issue go away.

**Comment by michaelklishin** [comment, marked as answer]
@annieblomgren can you please try this alpha build? https://github.com/rabbitmq/rabbitmq-server-binaries-dev/releases/tag/v3.8.21-alpha.2

**Reply by michaelklishin** [reply]
@carlhoerberg

**Reply by michaelklishin** [reply]
We are ready to produce `3.8.21` but it would be nice to make sure that we are not the only ones who see a behavior change after https://github.com/rabbitmq/rabbitmq-server/commit/e63d89a09d82730021aacf551e387fa4240f6245

**Reply by johanrhodin** [reply]
We'll take it for a spin during the next 24h

**Reply by johanrhodin** [reply]
I haven't been able to reproduce with `RabbitMQ 3.8.21-alpha.2` (was able with 3.8.18). 
Have a long running job doing this over and over for a few hours running now.

**Reply by annieblomgren** [reply]
yeah @michaelklishin can't reproduce it on v3.8.21-alpha.2 👍 
thank you for the fast response.

**Reply by lhoguin** [reply]
Thanks for the assist!

**Comment by michaelklishin** [comment]
[`3.9.3`](https://github.com/rabbitmq/rabbitmq-server/releases/tag/v3.9.3) and [`3.8.21`](https://github.com/rabbitmq/rabbitmq-server/releases/tag/v3.8.21) are released.
